### PR TITLE
Fix URI usage with user namespace

### LINKED ISSUE #2588: Running an image from docker in 3.0.x compared to 2.6.1: non-suid
## Version of Singularity

3.0.x
### Expected behavior

With singularity <3.0.x it was possible to exec an image from docker using without setuid using userNS both with overlay and with underlay it was also possible to use directories non existing in the image. For example using a standard alpine image and binding $PWD:/minnie where $PWD ≠ $HOME I can run a Hello World! script from the host node using overlay allegedly using a NON-SETUID workflow

-bash-4.2$ singularity --version
**2.6.1-dist**
-bash-4.2$ cd /pippo/aforti
-bash-4.2$ cat /proc/sys/user/max_user_namespaces 
15000
-bash-4.2$  egrep 'setuid|lay ' /etc/singularity/singularity.conf|grep -v '#'
**allow setuid = no
enable overlay = yes**
enable underlay = yes
-bash-4.2$ lsb_release -r
Release:	7.6.1810
-bash-4.2$ uname -r
3.10.0-957.1.3.el7.x86_64
-bash-4.2$  singularity --verbose exec -C -B $PWD:/minnie docker://alpine cat /minnie/pluto
Increasing verbosity level (2)
Singularity version: 2.6.1-dist
Exec'ing: /usr/libexec/singularity/cli/exec.exec
Evaluating args: '-C -B /pippo/aforti:/minnie docker://alpine cat /minnie/pluto'
VERBOSE: Set messagelevel to: 2
VERBOSE: Initialize configuration file: /etc/singularity/singularity.conf
VERBOSE Docker image: alpine
VERBOSE Registry: index.docker.io
VERBOSE Namespace: library
VERBOSE Repo Name: alpine
VERBOSE Repo Tag: latest
VERBOSE Version: None
VERBOSE Obtaining tags: https://index.docker.io/v2/library/alpine/tags/list
Docker image path: index.docker.io/library/alpine:latest
VERBOSE Obtaining manifest: https://index.docker.io/v2/library/alpine/manifests/latest
VERBOSE Obtaining manifest: https://index.docker.io/v2/library/alpine/manifests/latest
VERBOSE Obtaining manifest: https://index.docker.io/v2/library/alpine/manifests/sha256:3d2e482b82608d153a374df3357c0291589a61cc194ec4a9ca2381073a17f58e
Cache folder set to /home/aforti/.singularity/docker
Creating container runtime...
Importing: base Singularity environment
Exploding layer: sha256:cd784148e3483c2c86c50a48e535302ab0288bebd587accf40b714fffd0646b3.tar.gz
VERBOSE: Set messagelevel to: 2
VERBOSE: Initializing Singularity Registry
Exploding layer: sha256:1f6d219d0077f6af480398f264117d2e74c03ec5f12d09b7b29b1eefbba872eb.tar.gz
VERBOSE: Set messagelevel to: 2
VERBOSE: Initializing Singularity Registry
VERBOSE: Set messagelevel to: 2
VERBOSE: Initialize configuration file: /etc/singularity/singularity.conf
VERBOSE: Initializing Singularity Registry
VERBOSE: Invoking non-SUID program flow: /usr/libexec/singularity/bin/action
VERBOSE: Set messagelevel to: 2
VERBOSE: Initialize configuration file: /etc/singularity/singularity.conf
VERBOSE: Initializing Singularity Registry
**VERBOSE: Running NON-SUID program workflow**
VERBOSE: Invoking the user namespace
VERBOSE: No autofs bug path in configuration, skipping
VERBOSE: Exec'ing cleanupd thread: /usr/libexec/singularity/bin/cleanupd
VERBOSE: Set messagelevel to: 2
VERBOSE: Daemonizing cleandir cleanup process
VERBOSE: Using session directory: /var/singularity/mnt/session
VERBOSE: Mounting overlay with options: lowerdir=/var/singularity/mnt/container,upperdir=/var/singularity/mnt/overlay/upper,workdir=/var/singularity/mnt/overlay/work
VERBOSE: Singularity overlay mount did not work (Operation not permitted), continuing without it
VERBOSE: Running all mount components
VERBOSE: Couldn't mount devpts filesystem, continuing with PTY functionality disabled.
VERBOSE: Queuing mount of new procfs
VERBOSE: Queuing bind mount of /sys
VERBOSE: Using sessiondir for home directory
VERBOSE: Queuing bind mount of staged home directory to container: /var/singularity/mnt/session//home/aforti -> /home/aforti
VERBOSE: Queuing bind mount of '/pippo/aforti' to '/minnie'
VERBOSE: Queuing bind mount of /tmp
VERBOSE: Queuing bind mount of /var/tmp
VERBOSE: Not mounting current directory: contain was requested
VERBOSE: Mounting '/var/singularity/mnt/session/dev' at '/var/singularity/mnt/final/dev'
VERBOSE: Mounting 'proc' at '/var/singularity/mnt/final/proc'
VERBOSE: Mounting '/sys' at '/var/singularity/mnt/final/sys'
VERBOSE: Mounting '/var/singularity/mnt/session/home/aforti' at '/var/singularity/mnt/final/home/aforti'
VERBOSE: Mounting '/pippo/aforti' at '/var/singularity/mnt/final/minnie'
VERBOSE: Mounting '/var/singularity/mnt/session/tmp' at '/var/singularity/mnt/final/tmp'
VERBOSE: Mounting '/var/singularity/mnt/session/var_tmp' at '/var/singularity/mnt/final/var/tmp'
VERBOSE: Running file components
VERBOSE: Creating template passwd file and appending user data: /var/singularity/mnt/session/passwd
VERBOSE: Binding file '/var/singularity/mnt/session/passwd' to '/var/singularity/mnt/final/etc/passwd'
VERBOSE: Updating group file with user info
VERBOSE: Binding file '/var/singularity/mnt/session/group' to '/var/singularity/mnt/final/etc/group'
VERBOSE: Binding file '/etc/resolv.conf' to '/var/singularity/mnt/final/etc/resolv.conf'
VERBOSE: Containing all rootfs components
VERBOSE: Entering container file system root: /var/singularity/mnt/final
LOG    : USER=aforti, IMAGE='alpine', COMMAND='exec'
VERBOSE: Exec'ing /.singularity.d/actions/exec
**Hello World!**
VERBOSE: Cleaning directory: /tmp/.singularity-runtime.19KgOs0H

### Actual behavior
If I use singularity 3.0.3 instead. With the same setup it fails

-bash-4.2$ singularity --version
**singularity version 3.0.3-rc2.el7**
-bash-4.2$ cd /pippo/aforti
-bash-4.2$ cat /proc/sys/user/max_user_namespaces 
15000
-bash-4.2$ egrep 'setuid|lay ' /etc/singularity/singularity.conf|grep -v '#'
allow setuid = no
enable overlay = yes
enable underlay = yes
-bash-4.2$ lsb_release -r
Release:	7.6.1810
-bash-4.2$ uname -r
3.10.0-957.1.3.el7.x86_64
-bash-4.2$ singularity --verbose exec -C -B $PWD:/minnie docker://alpine cat /minnie/pluto
VERBOSE: Set messagelevel to: 4
VERBOSE: Container runtime
VERBOSE: Check if we are running as setuid
VERBOSE: Spawn scontainer stage 1
VERBOSE: Execute scontainer stage 1
**VERBOSE: Create user namespace**
VERBOSE: Create mount namespace
VERBOSE: Create pid namespace
VERBOSE: Spawn smaster process
VERBOSE: Spawn scontainer stage 2
VERBOSE: Create ipc namespace
VERBOSE: Create mount namespace
VERBOSE: Spawn RPC server
VERBOSE: Execute smaster process
VERBOSE: Serve RPC requests
VERBOSE: Not mounting current directory: container was requested
**FATAL:   container creation failed: mount error: can't mount image /proc/self/fd/9: failed to find loop device: could not attach image file too loop device: permission denied**

Note that this is not only a overlay issue. Underlay doesn't work either

### Steps to reproduce behavior

steps above.

Is this expected? or is it a bug in 3.0.x. Running user images from docker is an important workflow for us (and others). If we cannot run like we do in 2.6.1 we cannot use non-suid mode.

@lukasheinrich

### PR DESCRIPTION
**Pull request by cclerget**
**Description of the Pull Request (PR):**

This PR fixes a regression from v2 allowing user namespace to execute container directly from URI like :

```
singularity shell -u docker://busybox
```

**This fixes or addresses the following GitHub issues:**

- Fixes #2588

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: internal/pkg/runtime/engines/singularity/cleanup.go [hunk 1]
```diff
@@ -25,6 +25,13 @@ import (
 func (engine *EngineOperations) CleanupContainer(fatal error, status syscall.WaitStatus) error {
 	sylog.Debugf("Cleanup container")
 
+	if engine.EngineConfig.GetDeleteImage() {
+		image := engine.EngineConfig.GetImage()
+		if err := os.RemoveAll(image); err != nil {
+			sylog.Errorf("failed to delete container image %s: %s", image, err)
```
Comment (bauerm97) [inline internal/pkg/runtime/engines/singularity/cleanup.go:31]: Are we intentionally calling `Errorf` and not exiting here? This makes it possible to fail the image deletion and still return a 0 exit code

File: internal/pkg/runtime/engines/singularity/config/config.go [hunk 2]
```diff
@@ -90,6 +90,7 @@ type JSONConfig struct {
 	TargetUID     int           `json:"targetUID,omitempty"`
 	TargetGID     []int         `json:"targetGID,omitempty"`
 	LibrariesPath []string      `json:"librariesPath,omitempty"`
+	DeleteImage   bool          `json:"deleteImage,omitempty"`
```
Comment (bauerm97) [inline internal/pkg/runtime/engines/singularity/config/config.go:93]: With this option, doesn't this render it impossible for two executions of singularity to share the same temporary splatted fs?

File: cmd/internal/cli/actions_linux.go [hunk 3]
```diff
@@ -30,9 +34,62 @@ import (
...
+		tmpdir = filepath.Join(pw.Dir, ".singularity", "tmp")
+		if !fs.IsDir(tmpdir) {
+			if err := os.Mkdir(tmpdir, 0755); err != nil {
+				return "", fmt.Errorf("could not create directory %s: %s", tmpdir, err)
+			}
+		}
+	}
+
+	// create temporary sandbox
+	dir, err := ioutil.TempDir(tmpdir, "rootfs-")
```
Comment (bauerm97) [inline cmd/internal/cli/actions_linux.go:82]: I'm questioning this approach. It should be possible to have a shared cache of splatted out images (I believe this was the behavior in 2.x, correct me if I'm wrong anyone). Even if this isn't the case with 2.x, I still think it should be possible:

1. Aquire shared lock on a file representing the hash of the fs we want to splat
2. If we're first to aquire this lock (i.e. create this file) then splat the image out to disk
3. After image extraction is complete, write a second file (or write some data to hash file) to signal that splatted image can be used
4. Aquire exclusive lock on file (block other files from opening/shared locking)
5. Check if any other shared locks on this file exist
6. Delete splatted filesystem if no other shared locks exist
7. Release both exclusive and shared lock

The approach outlined above does seem much more complex, so perhaps we should merge this Pull Request as is and discuss the possibility of a shared cache and revisit if applicable

**Review by bauerm97 — CHANGES_REQUESTED** [review]
Just some general comments from a quick read through while on the airplane. Once I land in San Diego I'll try to revisit and see if there is anything which needs changing.

One thing I would like to see is the inclusion of tests here so we don't get another regression down the line

File: internal/pkg/runtime/engines/singularity/cleanup.go (hunk 1 above)
Comment (cclerget) [reply internal/pkg/runtime/engines/singularity/cleanup.go:31]: Yes it's best effort, we can't return error here and interrupt all other cleanup steps like cgroups, network and let garbage on node

File: internal/pkg/runtime/engines/singularity/config/config.go (hunk 2 above)
Comment (cclerget) [reply internal/pkg/runtime/engines/singularity/config/config.go:93]: each execution has its own copy

File: cmd/internal/cli/actions_linux.go (hunk 3 above)
Comment (cclerget) [reply cmd/internal/cli/actions_linux.go:82]: The PR approach is strictly identical to 2.x, create a temporary folder and delete it after container execution, those images were not cached in 2.x.

Now if we go this way, imo it should be handled by cache package instead and additionally deal with multi-node cases with the concern that some network filesystem doesn't support file locking

**Review by bauerm97 — APPROVED** [review]

**Pull request merged**